sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/HadoopFileLinesReader.scala

```diff
@@ -42,7 +45,10 @@ class HadoopFileLinesReader(
       Array.empty)
     val attemptId = new TaskAttemptID(new TaskID(new JobID(), TaskType.MAP, 0), 0)
     val hadoopAttemptContext = new TaskAttemptContextImpl(conf, attemptId)
-    val reader = new LineRecordReader()
+    val reader = recordDelimiter match {
+      case Some(delim) => new LineRecordReader(delim)
+      case _ => new LineRecordReader()
```

Comment:
The problem here is, we can't cover only when `charset` is set. It covers \r \r\n and \n in UTF-8 by default. If charset alone is given, it should also cover this case likewise.

BTW, I am leaving all those comments because I have walked through the almost identical change a long ago in my local and I just decided to not propose that change.

Reply:
One more note: this was the end of my try. I just gave up with a conclusion that we need a complete rewrite of `LineRecordReader` and thought it's not worth.

Reply:
how about we force users to provide a customer recordDelimiter if charset is specified?

Reply:
I believe arguably the more common case is a case when `charset` is set alone. It looks a bit odds that `charset` option is dependent on `recordDelimiter`.
If I understood correctly,`charset` can't be used alone in that way but I believe that's what most users would try because the same option `charset` is there in CSV.

Reply:
How does CSV solve this problem?

Reply:
CSV has the same issue. That option is incomplete. So, I believe we currently support ascii-compatible encodings only in CSV.

Reply:
For example, see:

```
curl -O https://raw.githubusercontent.com/HyukjinKwon/spark/264a1dc603164bd264e0c084608f31ffb8ad5f69/sql/core/src/test/resources/cars_utf-16.csv
```

```
scala> spark.read.option("encoding", "utf-16").option("header", true).csv("cars_utf-16.csv").show()
+----+-----+-----+--------------------+------+
|year| make|model|             comment|blank�|
+----+-----+-----+--------------------+------+
|2012|Tesla|    S|          No comment|     �|
|   �| null| null|                null|  null|
|1997| Ford| E350|Go get one now th...|     �|
|2015|Chevy|Volt�|                null|  null|
+----+-----+-----+--------------------+------+
```

So, I was thinking we should rather deprecate this option in CSV.

Reply:
I had an offline discussion with @cloud-fan about https://github.com/MaxGekk/spark-1/pull/1#discussion_r176963920. I am fine with that. There seems at least no hole.